### PR TITLE
Allow to pass a dictionary of preloaded textures

### LINKED ISSUE #10: Option for python interface with preloaded images
Have a dict:

```
{
    "texturePath0": <FileLike>,
    "texturePath1": <FileLike>,
}
```

### PR DESCRIPTION
**Pull request by Breush**
Closes #10

### CODE REVIEW

File: yoga/model/helpers.py [hunk 1]
```diff
@@ -1,11 +1,61 @@
 from ._assimp import ffi
 
 import io
+import re
 import os.path
+import unidecode
```
Comment (flozz) [inline yoga/model/helpers.py]: Please add a blank line before and after this import

```
import std libs

import pypi libs

import project modules
```

File: requirements.txt [hunk 2]
```diff
@@ -32,5 +32,6 @@ Sphinx==1.7.1
 sphinx-rtd-theme==0.2.4
 sphinxcontrib-websupport==1.0.1
 typing==3.6.4
+unidecode==1.0.22
```
Comment (flozz) [inline requirements.txt]: you have to add the dependency to the `setup.py` file too ;)

File: test/test_model_helpers.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# coding=utf8
```
Comment (flozz) [inline test/test_model_helpers.py]: this work but the `coding: utf-8` form is more wildly used :)

File: yoga/model/helpers.py [hunk 4]
```diff
@@ -1,11 +1,61 @@
...
+    # Expects a unicode path, returns a ascii one.
+    # Paths are normalized to a standard linux relative path,
+    # without a point, and lowercase.
+    # That is to say /images\subfolder/..\texture.png -> images/texture.png
+    # It does not correspond to an effective path,
+    # as the backslashes on linux are wrongly seen as separators.
+    # This function is meant to give a standard output.
+
+    path = unidecode.unidecode(path)
+    split_path = re.findall(r"[\w\s\-_.:]+", path)
```
Comment (flozz) [inline yoga/model/helpers.py]: why not

```python
splited_path = path.replace("\\", "/").split("/")
```

File: yoga/model/helpers.py [hunk 5]
```diff
@@ -1,11 +1,61 @@
...
+        if name == "." or name[-1:] == ":":
+            continue
+        elif name == "..":
+            ignored_folders += 1
+        elif ignored_folders > 0:
+            ignored_folders -= 1
+        elif i == 0:
+            normalized_path = name
+        else:
+            normalized_path = name + "/" + normalized_path
```
Comment (flozz) [inline yoga/model/helpers.py]: maybe you can use a list here ?

```python
normalized_path = []
...
 normalized_path.push(name)
...
return "/".join(normalized_path[::-1]).lower()

```

File: yoga/model/helpers.py [hunk 6]
```diff
@@ -1,11 +1,61 @@
...
+    # It does not correspond to an effective path,
+    # as the backslashes on linux are wrongly seen as separators.
+    # This function is meant to give a standard output.
+
+    path = unidecode.unidecode(path)
+    split_path = re.findall(r"[\w\s\-_.:]+", path)
+    normalized_path = ""
+    ignored_folders = 0
+
+    for i, name in enumerate(reversed(split_path)):
```
Comment (flozz) [inline yoga/model/helpers.py]: the reverse can be done using slices:
`reversed(split_path)` → `split_path[::-1]`

you can also doing it earlier like this:

```python
splited_path = path.replace("\\", "/").split("/")[::-1]
```

File: yoga/model/helpers.py [hunk 7]
```diff
@@ -37,76 +87,73 @@ def find_valid_path(path, root_path):
...
 
 def find_valid_texture_path(path, textures):
-    # In textures, all paths are supposed to be relative
+    # The path and the textures' paths are supposed to have
+    # already been normalized.
 
-    tested_path = path
-    if tested_path in textures:
-        return tested_path
+    split_path = reversed(path.split("/"))
```
Comment (flozz) [inline yoga/model/helpers.py]: does it work wi windows path too (`\`) ?

Comment (flozz) [reply yoga/model/helpers.py]: also, you can use slices to reverse the list (`[::-1]`) :)

File: requirements.txt (hunk 2 above)
Comment (Breush) [reply requirements.txt]: Done

**Pull request merged**